### PR TITLE
Author datatype, close #29

### LINKED ISSUE #29: Author identification?
I wish that authors can be identified by not just their names but also (optionally) their ORCID, ResearcherID, DAI, etc. This would require changes to the metadata format in `meta_data.lean`.

### PR DESCRIPTION
**Pull request by andrejbauer**
I apologize for the merge.

### CODE REVIEW

File: fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean [hunk 1]
```diff
@@ -61,14 +62,34 @@ unfinished Kepler_conjecture :
 { description := "Proof of Kepler conjecture",
   references := [cite.Item cite.Ibidem "Theorem X.Y"] }
 
-open result
-
 def fabstract : meta_data := {
   description := "This article announces the formal proof of the Kepler conjecture on dense sphere packings in a combination of the HOL Light and Isabelle/HOL proof assistants.  It represents the primary result of the now completed Flyspeck project.",
-  authors := ["Thomas Hales", "Mark Adams", "Gertrud Bauer", "Tat Dat Dang", "John Harrison", "Le Truong Hoang", "Cezary Kaliszyk", "Victor Magron", "Sean McLaughlin", "Tat Thang Nguyen", "Quang Truong Nguyen", "Tobias Nipkow", "Steven Obua", "Joseph Pleso", "Jason Rute", "Alexey Solovyev", "Thi Hoai An Ta", "Nam Trung Tran", "Thi Diep Trieu", "Josef Urban", "Ky Vu", "Roland Zumkeller"],
+  authors := [
+    ({name := "Thomas Hales"} : author), -- TODO why do I have cast the type here but not in other fabstracts?
```
Comment (andrejbauer) [inline fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: I think one of those `open` statements must be messing with the type inference. In other fabstracts no such type cast was needed. And I tried `{| author, name := "Thomas Hales" |}` but it did not work. These complications make me doubt existence of Thomas Hales once again.

Comment (robertylewis) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: Hmm, I'm looking into it. But a slightly cleaner workaround, for now, is
```lean
authors := [
    {author . name := "Thomas Hales"}, ...
```

Comment (favonia) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: Question: are the spaces around the dot crucial?

Comment (robertylewis) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: There has to be a space after the dot, but `{author. name := "Thomas Hales"}` is accepted.

Comment (robertylewis) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: This is a strange bug. It appears to be related to the order in which dependencies are imported. It's nondeterministic, or at least dependent on superficial changes, like spaces in the `import` lines.

Comment (andrejbauer) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: Presumably we don't have to remove the bug in Lean before we accept this PR? Any other comments about the PR?

Comment (Kha) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: I investigated the bug, it's due to combining the list/vector-overloaded `[...]` notation with anonymous structure instances. I have a tentative local fix, but as you said it should be okay to merge this PR as is.

**Comment by robertylewis** [comment]
I can't use the "rebase and merge" button due to conflicts, probably from the merge commit. I'm going to try "squash and merge." Hopefully that doesn't make a new merge commit.

File: fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean (hunk 1 above)
Comment (robertylewis) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: Thanks @Kha !

Comment (Kha) [reply fabstract/Hales_T_et_al_Kepler_conj/fabstract.lean:68]: https://github.com/leanprover/lean/pull/1754 has been merged

**Pull request merged**